### PR TITLE
Add selector-no-empty rule

### LINKED ISSUE #1777: Add selector-no-empty rule
Are syntax error in scope of this linter? I'd imagine a rule that catches trailing commas in selector lists like

```css
.a, .b, {
  color: red;
}
```
would be quite useful. Here's a [recent case](https://github.com/StylishThemes/GitHub-Dark/commit/e5bf405f2af578c5feed849893c56dc575deb37d).

### PR DESCRIPTION
**Pull request by gaidarenko**
Closes #1777

### CODE REVIEW

**Comment by evilebottnawi** [comment]
@gaidarenko add note to `CHANGELOG.md`

**Comment by evilebottnawi** [comment]
@gaidarenko oh sorry, we add this manually after merge, it is not necessary to do so

**Comment by evilebottnawi** [comment]
@gaidarenko and next PR please use template, thanks :smile:

File: src/rules/selector-no-empty/README.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
+# selector-no-empty
+
+Disallow empty selectors.
+
+```css
```
Comment (jeddy3) [inline src/rules/selector-no-empty/README.md]:
```css
  a, , b {}
/** ↑ 
 * An empty selector */
```

File: src/rules/selector-no-empty/README.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    a { color: pink; }
+/** ↑ 
+ * This is selector */
+```
+
+Empty selectors (by themselves or within a selector list) invalidate a rule. This rule splits the selector list on, and any part of zero length or contains only whitespace is consider empty. Everything else isn't.
```
Comment (jeddy3) [inline src/rules/selector-no-empty/README.md]: "Empty selectors (by themselves or within a selector list) invalidate a rule. This rule splits selector lists on `,`. Any part of zero length or containing only whitespace is deemed empty."

File: src/rules/selector-no-empty/README.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+## Options
+
+### `true`
+
+The following patterns are considered warnings:
+
+```css
+{ color: red} /* entirely empty selector */
```
Comment (jeddy3) [inline src/rules/selector-no-empty/README.md]: Please remove all the declarations from within the blocks for consistency with the other `selector-*` rule READMEs.

I think we can also remove the comments as I think the examples are pretty self-explanatory.

File: src/rules/selector-no-empty/README.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+```css
+a { color: pink; }
+```
+
+```css
+a, b { color: pink; }
+```
+
+```css
```
Comment (jeddy3) [inline src/rules/selector-no-empty/README.md]: Please remove this example as it's not connected to the preceding patterns.

File: src/rules/selector-no-empty/__tests__/index.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+import { testRule } from "../../../testUtils"
+
+const rule = rules[ruleName]
+
+testRule(rule, {
+  ruleName,
+  config: [true],
+
+  accept: [ {
+    code: ".a, .b {color: red}",
```
Comment (jeddy3) [inline src/rules/selector-no-empty/__tests__/index.js]: Please remove the declarations from within the blocks. The focus of the tests is the selector.

File: src/rules/selector-no-empty/__tests__/index.js [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    message: messages.rejected,
+    line: 1,
+    column: 11,
+  }, {
+    code: "   .a,    .b, {}",
+    message: messages.rejected,
+    line: 1,
+    column: 14,
+  } ],
+})
```
Comment (jeddy3) [inline src/rules/selector-no-empty/__tests__/index.js]: Can we get one or two tests for the other parsers please e.g. `sugarss`, `less` and `scss`? I'm curious to seem, especially `sugarss` parse empty selectors in the same way.

**Comment by jeddy3** [comment]
@gaidarenko Thanks for tackling yet another rule! You've been doing a great job lately.

I've requested some documentation and test changes.

**Comment by gaidarenko** [comment]
Done

File: src/rules/selector-no-empty/README.md [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+Disallow empty selectors.
+
+```css
+  a, , b {}
+/** ↑ 
+ * An empty selector */
+```
+
+Empty selectors (by themselves or within a selector list) invalidate a rule. This rule splits the selector list on `,`. Any part of zero length or containing only whitespace is deemed empty.
```
Comment (davidtheclark) [inline src/rules/selector-no-empty/README.md]: > Any part of zero length

?? How about "Any part that has zero length"

File: src/rules/selector-no-empty/__tests__/index.js [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+    message: messages.rejected,
+    line: 1,
+    column: 4,
+  }, {
+    code: "   ,  .a   , .b {}",
+    message: messages.rejected,
+    line: 1,
+    column: 4,
+  }, {
+    code: ".a,  ,.b {}",
```
Comment (davidtheclark) [inline src/rules/selector-no-empty/__tests__/index.js]: We should have a test for the zero-length selector mentioned in the readme.

```css
a,,b {}
```

File: src/rules/selector-no-empty/__tests__/index.js [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+    code: " a {}",
+  } ],
+
+  reject: [ {
+    code: "   {}",
+    message: messages.rejected,
+    line: 1,
+    column: 4,
+  }, {
+    code: "   ,  .a   , .b {}",
```
Comment (davidtheclark) [inline src/rules/selector-no-empty/__tests__/index.js]: How about 

```css
  .a      ,    , .b {}
```

I'm curious if the column count stays accurate.

File: src/rules/selector-no-empty/index.js [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    const validOptions = validateOptions(result, ruleName, { actual })
+    if (!validOptions) { return }
+
+    root.walkRules(rule => {
+      let index = 0
+
+      rule.selector.split(",").forEach(item => {
+        index += item.length + 1
+
+        if (item.trim()) { return }
```
Comment (davidtheclark) [inline src/rules/selector-no-empty/index.js]: Let's be explicit that `item.trim() !== ''`. That will make the intention more clear.

File: src/rules/selector-no-empty/README.md (hunk 7 above)
Comment (gaidarenko) [inline src/rules/selector-no-empty/README.md]: @davidtheclark I think you should discuss this issue with @jeddy3, because it was his idea.

**Comment by gaidarenko** [comment]
Done.

File: src/rules/selector-no-empty/README.md (hunk 7 above)
Comment (gaidarenko) [inline src/rules/selector-no-empty/README.md]: Okay

**Comment by gaidarenko** [comment]
Wording changed.

**Comment by davidtheclark** [comment]
LGTM, thanks! @jeddy3, ready to merge?

**Comment by jeddy3** [comment]
@davidtheclark Yep! LGTM.

@gaidarenko Thanks for ticking off another rule :)

**Pull request merged**